Fix reported amount on incoming transactions carrying withdrawals

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1952 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- d1c5618b79ac861600e94b7200a94c89a7b80f87
  :round_pushpin: **modify scenario spending rewards to send money to _another_ wallet and assess right amount**
    We didn't quite catch a bug here because we were only looking at one side of the coin by sending the transaction to ourselves (so, we couldn't really assess the amount on an _incoming_ transaction, from a different perspective. That is now done properly such that 1 lovelace is send to another wallet (and, which should be the amount reported by this wallet). The outgoing amount however should contain any output sent to _another_ wallet, as well as fees.

- 6932d1d7d73677d840a5cebc7ac96d31eca96f42
  :round_pushpin: **revise 'amount' calculation in the primitive model**
    Such that we only count withdrawals in outgoing transactions. In principle, withdrawals should be necessarily ours on a transaction.. but it's not excluded that a transaction may contain multiple withdrawals from different wallets (especially in the context of multisig). Beside, withdrawals should be completely out of the picture for _incoming_ transactions.

- aa9fdf9ba996898a6978a4eb4af638361e466bcc
  :round_pushpin: **add test showing oddities on total balance + rewards when there are pending withdrawals + fix it**
    This was trickier than I originally thought because crafting a generator of pending transactions which is somewhat meaningful and can have withdrawals turned out to be a bit complex. The property originally failed (without the fix) and clearly showed the bug described in #1955. The fix is relatively straightforward and is about adding the reward amount only if there are no pending withdrawal

- 5867d38f818e14c802de8293aed1ac425227e478
  :round_pushpin: **fix SlotNo generator in database arbitrary instances**
    This one was wrongly changed to browse the entire Word32 space instead of staying within small bounds as originally. There are many things in tests and other generators derived from the slot number value, and too big values can really do bad things here.

- 8d1eadb42cc263116f27b625a8cb5438653b63f9
  :round_pushpin: **prevent users from creating two withdrawals simultaneously on the same wallet**
    Rewards ought to be spent fully, therefore, making two concurrent withdrawals has no sense since only one transaction will eventually be validated.


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
modify scenario spending rewards to send money to _another_ wallet and assess right amount

  We didn't quite catch a bug here because we were only looking at one side of the coin by sending the transaction to ourselves (so, we couldn't really assess the amount on an _incoming_ transaction, from a different perspective. That is now done properly such that 1 lovelace is send to another wallet (and, which should be the amount reported by this wallet). The outgoing amount however should contain any output sent to _another_ wallet, as well as fees.

KtorZ committed:
revise 'amount' calculation in the primitive model

  Such that we only count withdrawals in outgoing transactions. In principle, withdrawals should be necessarily ours on a transaction.. but it's not excluded that a transaction may contain multiple withdrawals from different wallets (especially in the context of multisig). Beside, withdrawals should be completely out of the picture for _incoming_ transactions.

paweljakubas commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -244,30 +245,38 @@ spec = do
                 Default Empty
             verify rw2 [ expectListSize 1 ]
 
+        eventually "There's one incoming transaction with correct amount" $ do
```
:+1:

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -369,27 +376,25 @@ prefilterBlock b u0 = runState $ do
-        let received = fromIntegral @_ @Integer $ balance ourU
-        let spent = fromIntegral @_ @Integer $ balance (u `restrictedBy` ourIns)
-        let amt = fromIntegral $ abs (received - spent - withdrawalAmt)
+        ourWithdrawals <- mapMaybeM ourChimericAccount $ Map.toList $ withdrawals tx
```
yes, we need to check if withdrawals are ours

paweljakubas reviewed (approved):
looks like a fix! extension to testing persuasive

Anviking reviewed (approved):
Fix looks good to me. Wanted to check that the adjusted tests cover the fix, but I only get socket errors  when I try to run the tests locally…

piotr-iohk reviewed (approved):
Looks good also testing manually.

KtorZ committed:
add test showing oddities on total balance + rewards when there are pending withdrawals + fix it

  This was trickier than I originally thought because crafting a generator of pending transactions which is somewhat meaningful and can have withdrawals turned out to be a bit complex. The property originally failed (without the fix) and clearly showed the bug described in #1955. The fix is relatively straightforward and is about adding the reward amount only if there are no pending withdrawal

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs:
```diff
@@ -214,6 +229,40 @@ prop_initialBlockHeight s =
+    $ classify hasWithdrawals "has withdrawals"
+    $ if hasPending
+      then property (totalWithPending < totalWithoutPending)
+      else property (totalWithPending == totalWithoutPending)
```
The property is actually more precise than it says "counting rewards once". It really shows that having pending transactions result in a smaller balance for the wallet, regardless of whether there are withdrawals or not. 

Of course, this only holds in the presence of fees that are non-null.

KtorZ replied:
```
only counts rewards once.
      +++ OK, passed 10000 tests:
      45.33% has pending
       5.21% has withdrawals
```

Anviking replied:
👍

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -641,8 +642,8 @@ mkShelleyWallet ctx wid cp meta pending progress = do
             { available = Quantity $ availableBalance pending cp
-            , total = Quantity $ reward + totalBalance pending cp
-            , reward = Quantity reward
+            , total = Quantity $ totalBalance pending reward cp
```
👍

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -289,14 +289,23 @@ availableBalance :: Set Tx -> Wallet s -> Natural
-totalBalance pending =
-    balance . totalUTxO pending
+totalBalance pending (Quantity rewards) s =
+    balance (totalUTxO pending s) + if hasPendingWithdrawals then 0 else rewards
```
👍

KtorZ committed:
fix SlotNo generator in database arbitrary instances

  This one was wrongly changed to browse the entire Word32 space instead of staying within small bounds as originally. There are many things in tests and other generators derived from the slot number value, and too big values can really do bad things here.

KtorZ committed:
prevent users from creating two withdrawals simultaneously on the same wallet

  Rewards ought to be spent fully, therefore, making two concurrent withdrawals has no sense since only one transaction will eventually be validated.
